src/server/worker/datum/create.go

```diff
@@ -0,0 +1,482 @@
+				BaseIndex: createBaseIndex(int64(i)),
+			})
+			if err != nil {
+				return err
+			}
+			inputs = append(inputs, input)
+		}
+		resultFileSetIDs := make([]string, len(inputs))
+		if err := task.DoBatch(ctx, taskDoer, inputs, func(i int64, output *types.Any, err error) error {
+			if err != nil {
+				return err
+			}
+			result, err := deserializePFSTaskResult(output)
+			if err != nil {
+				return err
+			}
+			if err := renewer.Add(ctx, result.FileSetId); err != nil {
+				return err
+			}
+			resultFileSetIDs[i] = result.FileSetId
+			return nil
+		}); err != nil {
+			return err
+		}
+		outputFileSetID, err = ComposeFileSets(ctx, taskDoer, resultFileSetIDs)
+		return err
+	}); err != nil {
+		return "", err
+	}
+	return outputFileSetID, nil
+}
+
+func createBaseIndex(index int64) int64 {
+	return index * int64(math.Pow(float64(10), float64(16)))
+}
+
+func ComposeFileSets(ctx context.Context, taskDoer task.Doer, fileSetIDs []string) (string, error) {
+	input, err := serializeComposeTask(&ComposeTask{
+		FileSetIds: fileSetIDs,
+	})
+	if err != nil {
+		return "", err
+	}
+	output, err := task.DoOne(ctx, taskDoer, input)
+	if err != nil {
+		return "", err
+	}
+	result, err := deserializeComposeTaskResult(output)
+	if err != nil {
+		return "", err
+	}
+	return result.FileSetId, nil
+}
+
+func createUnion(pachClient *client.APIClient, taskDoer task.Doer, inputs []*pps.Input) (string, error) {
+	var outputFileSetID string
+	if err := pachClient.WithRenewer(func(ctx context.Context, renewer *renew.StringSet) error {
+		pachClient := pachClient.WithCtx(ctx)
+		var fileSetIDs []string
+		for _, input := range inputs {
+			fileSetID, err := Create(pachClient, taskDoer, input)
+			if err != nil {
+				return err
+			}
+			if err := renewer.Add(ctx, fileSetID); err != nil {
+				return err
+			}
+			fileSetIDs = append(fileSetIDs, fileSetID)
+		}
+		var err error
+		outputFileSetID, err = ComposeFileSets(ctx, taskDoer, fileSetIDs)
+		return err
+	}); err != nil {
+		return "", err
+	}
+	return outputFileSetID, nil
+}
+
+func createCross(pachClient *client.APIClient, taskDoer task.Doer, inputs []*pps.Input) (string, error) {
+	var outputFileSetID string
+	if err := pachClient.WithRenewer(func(ctx context.Context, renewer *renew.StringSet) error {
+		pachClient := pachClient.WithCtx(ctx)
+		var fileSetIDs []string
+		for _, input := range inputs {
+			fileSetID, err := Create(pachClient, taskDoer, input)
+			if err != nil {
+				return err
+			}
+			if err := renewer.Add(ctx, fileSetID); err != nil {
+				return err
+			}
+			fileSetIDs = append(fileSetIDs, fileSetID)
+		}
+		var maxIdx int
+		var baseFileSetID string
+		var baseFileSetShards []*pfs.PathRange
+		for i, fileSetID := range fileSetIDs {
+			shards, err := pachClient.ShardFileSet(fileSetID)
+			if err != nil {
+				return err
+			}
+			if len(shards) > len(baseFileSetShards) {
+				maxIdx = i
+				baseFileSetID = fileSetID
+				baseFileSetShards = shards
+			}
+		}
```

Comment:
should this just be added to `common.Shard`?

Reply:
This is slightly different, and I didn't want to make that utility more complicated. It has to differentiate between the file set whose shards were chosen and the other file sets.